services/private_ip_manager/src/main/java/com/futurewei/alcor/privateipmanager/repo/IpAddrRangeRepo.java

```diff
@@ -164,8 +166,12 @@ private IpAddrAlloc doAllocateIpAddr(String vpcId, int ipVersion, String ipAddr)
             }
 
             try {
+                CacheConfiguration cfg = new CacheConfiguration();
+                cfg.setName(getIpAddrCacheName(rangeId));
+                cfg.setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL);
+
                 ICache<String, IpAddrAlloc> ipAddrCache =
-                        cacheFactory.getCache(IpAddrAlloc.class, getIpAddrCacheName(rangeId));
+                        cacheFactory.getCache(IpAddrAlloc.class, cfg);
```

Comment:
@pkommoju We are hitting a IP allocation failure, likely causing by the change in this file. Could you try to reproduce this issue in your local and propose a solution (and make code directly to this branch)? Thanks.

More detailed logs from IP manager, failure at the end:
```
2021-08-20 08:12:44.053  INFO 32420 --- [nio-9004-exec-1] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.repo.IpAddrRangeRepo.createIpAddrRange() startTime: 808185834948500ns, endTime: 808186051074000ns, duration: 216ms
2021-08-20 08:12:44.054  INFO 32420 --- [nio-9004-exec-1] c.f.a.p.s.implement.IpAddrServiceImpl    : Create ip address range success, request: IpAddrRangeRequest{id='44a1b42c-8c74-4fed-a9e8-f0db3b804568', vpcId='9192a4d4-ffff-4ece-b3f0-8d36e3d88038', subnetId='9192a4d4-ffff-4ece-b3f0-8d36e3d88000', ipVersion=4, firstIp='10.0.0.1', lastIp='10.0.0.254', usedIps=0, totalIps=254}
2021-08-20 08:12:44.108  INFO 32420 --- [nio-9004-exec-1] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.service.implement.IpAddrServiceImpl.createIpAddrRange() startTime: 808185826488100ns, endTime: 808186105872800ns, duration: 279ms
2021-08-20 08:12:44.108  INFO 32420 --- [nio-9004-exec-1] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.controller.IpAddrController.createIpAddrRange() startTime: 808185806863500ns, endTime: 808186106032800ns, duration: 299ms
2021-08-20 08:12:44.199  INFO 32420 --- [nio-9004-exec-1] i.j.internal.reporters.LoggingReporter   : Span reported: 19c5df46e43f0868:119f40ed61423030:19c5df46e43f0868:1 - createIpAddrRange
2021-08-20 08:12:45.145  INFO 32420 --- [nio-9004-exec-2] global                                   : Getting or creating cache IpAddrCache-44a1b42c-8c74-4fed-a9e8-f0db3b804568 AtomicityMode is TRANSACTIONAL
2021-08-20 08:12:45.266  INFO 32420 --- [nio-9004-exec-2] global                                   : Retrieved cache IpAddrCache-44a1b42c-8c74-4fed-a9e8-f0db3b804568 AtomicityMode is TRANSACTIONAL
2021-08-20 08:12:45.441  INFO 32420 --- [nio-9004-exec-2] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.repo.IpAddrRangeRepo.allocateIpAddr() startTime: 808187127741900ns, endTime: 808187439280300ns, duration: 311ms
2021-08-20 08:12:45.441  INFO 32420 --- [nio-9004-exec-2] c.f.a.p.s.implement.IpAddrServiceImpl    : Allocate ip address success, request: IpAddrRequest{ipVersion=4, vpcId='9192a4d4-ffff-4ece-b3f0-8d36e3d88038', subnetId='9192a4d4-ffff-4ece-b3f0-8d36e3d88000', rangeId='44a1b42c-8c74-4fed-a9e8-f0db3b804568', ip='10.0.0.1', state='activated'}
2021-08-20 08:12:45.451  INFO 32420 --- [nio-9004-exec-2] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.service.implement.IpAddrServiceImpl.allocateIpAddr() startTime: 808187127625600ns, endTime: 808187449775400ns, duration: 322ms
2021-08-20 08:12:45.451  INFO 32420 --- [nio-9004-exec-2] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.controller.IpAddrController.allocateIpAddr() startTime: 808187127480000ns, endTime: 808187450186000ns, duration: 322ms
2021-08-20 08:12:45.459  INFO 32420 --- [nio-9004-exec-2] i.j.internal.reporters.LoggingReporter   : Span reported: f7abf3c5dd709f12:cdd48c47c05d4a75:1156c528c55d9674:1 - allocateIpAddr
2021-08-20 08:19:07.366  INFO 32420 --- [nio-9004-exec-5] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.repo.IpAddrRangeRepo.createIpAddrRange() startTime: 808569352538200ns, endTime: 808569364686900ns, duration: 12ms
2021-08-20 08:19:07.367  INFO 32420 --- [nio-9004-exec-5] c.f.a.p.s.implement.IpAddrServiceImpl    : Create ip address range success, request: IpAddrRangeRequest{id='c49181bb-4ad0-4cdf-a240-3443fa00df65', vpcId='9192a4d4-ffff-4ece-b3f0-8d36e3d88038', subnetId='9192a4d4-ffff-4ece-b3f0-8d36e3d88001', ipVersion=4, firstIp='10.0.1.1', lastIp='10.0.1.254', usedIps=0, totalIps=254}
2021-08-20 08:19:07.367  INFO 32420 --- [nio-9004-exec-5] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.service.implement.IpAddrServiceImpl.createIpAddrRange() startTime: 808569352499400ns, endTime: 808569365131900ns, duration: 12ms
2021-08-20 08:19:07.367  INFO 32420 --- [nio-9004-exec-5] c.f.alcor.common.stats.StatisticsAspect  : com.futurewei.alcor.privateipmanager.controller.IpAddrController.createIpAddrRange() startTime: 808569352336500ns, endTime: 808569365194600ns, duration: 12ms
2021-08-20 08:19:07.369  INFO 32420 --- [nio-9004-exec-5] i.j.internal.reporters.LoggingReporter   : Span reported: c854d7eac6e2fd40:1ff2cfc38a8d1790:c854d7eac6e2fd40:1 - createIpAddrRange
2021-08-20 08:19:07.418  INFO 32420 --- [nio-9004-exec-6] global                                   : Getting or creating cache IpAddrCache-44a1b42c-8c74-4fed-a9e8-f0db3b804568 AtomicityMode is TRANSACTIONAL
2021-08-20 08:19:07.420  INFO 32420 --- [nio-9004-exec-6] global                                   : Retrieved cache IpAddrCache-44a1b42c-8c74-4fed-a9e8-f0db3b804568 AtomicityMode is TRANSACTIONAL
2021-08-20 08:19:07.424  WARN 32420 --- [nio-9004-exec-6] c.f.a.p.repo.IpAddrRangeRepo             : Allocate ip address from 44a1b42c-8c74-4fed-a9e8-f0db3b804568 failed
2021-08-20 08:19:07.425  INFO 32420 --- [nio-9004-exec-6] global                                   : Getting or creating cache IpAddrCache-c49181bb-4ad0-4cdf-a240-3443fa00df65 AtomicityMode is TRANSACTIONAL

```

Reply:
I will work on it.

Reply:
> I will work on it.

Thanks. This issue was triggered when a subnet is created. SM=>PM=>Private ip manager to create a gateway port and allocate ip.

Reply:
I did build and run it  on Jenkins machine and it worked fine. I am doing a second run to make sure I didn't pick up wrong run environment.

Reply:
I am able to reproduce it. subnet got created but the IP address range in question (in the cache named IpAddrCache-e4efed2e-33ab-4783-a7ce-0254dfd695d2) is empty. Should there be something at the point of failure or is it legitimate condition that has to be dealt with is something I am going to debug and find.

Reply:
Set logging level in PIM to INFO and ping test succeeded. The IP address cache has three entries, but since this state is after the test finished, it is not clear the cache being is empty when the test failed is a legitimate state.
Since adding debug out made the test succeed, I am suspecting timing (race conditions). Need to further investigate.

Reply:
It ran fine with logging=INFO three times, and once without the change. I am rerunning the Jenkins job as is now. The nature of failure suggests that there may be more race conditions which have not been uncovered (or rather this failure has uncovered).